lucet-runtime/lucet-runtime-internals/src/instance/execution.rs

```diff
@@ -25,58 +28,68 @@
+//!     termination request.
 //! * `Guest function executing`
-//!   - terminable: `true`
+//!   - termination result: `Ok(KillSuccess::Signalled)`
 //!   - execution_domain: `Guest`
 //! * `Guest function returns`
-//!   - terminable: `true`
+//!   - termination result: `Ok(KillSuccess::Pending)`
 //!   - execution_domain: `Guest`
+//!   - If the guest function has already returned, the instance will be placed into the
+//!     `Terminated` domain, and will not run any guest the next time it is run.
 //! * `Hostcall called`
-//!   - terminable: `true`
+//!   - termination result: `Ok(KillSuccess::Pending)`
 //!   - execution_domain: `Hostcall`
 //! * `Hostcall executing`
-//!   - terminable: `true`
+//!   - termination result: `Ok(KillSuccess::Pending)` or `Err(KillError::NotTerminable)`
 //!   - execution_domain: `Hostcall, or Terminated`
 //!   - `execution_domain` will typically be `Hostcall`, but may be `Terminated` if termination of
-//!   the instance is requested during the hostcall.
-//!   - `terminable` will be false if and only if `execution_domain` is `Terminated`.
-//! * `Hostcall yields`
-//!   - This is a specific point in "Hostcall executing" and has no further semantics.
-//! * `Hostcall resumes`
-//!   - This is a specific point in "Hostcall executing" and has no further semantics.
+//!     the instance is requested during the hostcall.
+//!   - `KillSwitch::terminate` will return `NotTerminable` if and only if the instance has already
+//!     been terminated.
+//! * `Hostcall yields`, or `Hostcall resumes`
+//!   - These are specific points in "Hostcall executing" and has no further semantics.
 //! * `Hostcall returns`
-//!   - terminable: `true`
+//!   - termination result: `Ok(KillSuccess::Signalled)`
 //!   - execution_domain: `Guest`
-//!   - `execution_domain` may be `Terminated` before returning, in which case `terminable` will be
-//!   false, but the hostcall would then exit. If a hostcall successfully returns to its caller it
-//!   was not terminated, so the only state an instance will have after returning from a hostcall
-//!   will be that it's executing terminable guest code.
-
+//!   - `execution_domain` may be `Terminated` before returning, in which case `terminate` will
+//!     return `NotTerminable`, but the hostcall would then exit. If a hostcall successfully
+//!     returns to its caller it was not terminated, so the only state an instance will have after
+//!     returning from a hostcall will be that it is executing terminable guest code.
+//! * `Instance::reset` or `Instance dropped`
+//!   - termination result: `Err(KillError::Invalid)`
+//!   - execution_domain: `Pending`
+//!   - If an instance is reset, outstanding `KillSwitch` objects will no longer hold a valid
+//!     reference to the instance. In that case, calls to `terminate` will return `Invalid`, and
+//!     have no bearing on further execution.
```

Comment:
Added this because it's worth highlighting, and it's part of why we have the `Invalid` variant of our kill error.